models/source_list.py

```diff
@@ -53,10 +67,106 @@ class SourceList(Base, FileOnDiskMixin):
         )
     )
 
+    def __init__(self, *args, **kwargs):
+        FileOnDiskMixin.__init__(self, *args, **kwargs)
+        SeeChangeBase.__init__(self)  # don't pass kwargs as they could contain non-column key-values
+
+        self._data = None
+
+        # manually set all properties (columns or not)
+        for key, value in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, value)
+
+    @orm.reconstructor
+    def init_on_load(self):
+        Base.init_on_load(self)
+        FileOnDiskMixin.init_on_load(self)
+        self._data = None
+
+    def __repr__(self):
+        output = (
+            f'<SourceList(id={self.id}, '
+            f'image_id={self.image_id}, '
+            f'is_sub={self.is_sub}), '
+            f'num_sources= {self.num_sources}>'
+        )
+
+        return output
+
+    @property
+    def data(self):
+        """
+        The data in this source list. A table of sources and their properties.
+        """
+        if self._data is None and self.filepath is not None:
+            self._data = self.load()
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        # TODO: add astropy table?
+        if isinstance(value, pd.DataFrame):
+            value = value.to_records(index=False)
+
+        if not isinstance(value, np.ndarray) or value.dtype.names is None:
+            raise TypeError("data must be a pandas DataFrame or numpy recarray")
+
+        self._data = value
+        self.num_sources = len(value)
+
+    @property
+    def mjd(self):
+        """
+        The MJD of the image this source list was generated from.
+        """
+        if self.image is None:
+            return None
+        return self.image.mjd
+
+    def load(self):
+        """
+        Load this source list from the database.
+        """
+        # TODO: should we replace this with FITS and astropy tables?
+        return np.load(self.get_fullpath())  # this should always be a single file, right?
+
     def save(self):
         """
-        Save this source list to the database.
+        Save the data table to a file on disk.
         """
-        # TODO: Must implement this at some point!
-        self.filepath = uuid.uuid4().hex
+        if self.image is None:
+            raise ValueError("Cannot save source list without an image")
+
+        if self.data is None:
+            raise ValueError("Cannot save source list without data")
+
+        filename = self.invent_filename()
+
+        if self.is_sub:
+            filename += '_detections'
+        else:
+            filename += '_sources'
```

Comment:
I might just call it _sources always.  I don't think there's any benefit in having different filenames based on where the thing came from; at some level, the Source List class itself shouldn't care.

Reply:
But are they different?  They're the result of this algorithm run on an image.  There's nothing different about the algorithm, so what this process adds to the filename shouldn't vary.  You can tell a subtracted image vs. a regular image based on the rest of the filename.  (Subtracted images do have a different filename pattern than regular images, don't they?   I would argue that yes, they really ought to, for the reasons you say; but, once that happens, then we don't need the _detections vs _sources difference any more.)

Reply:
What happens if the naming convention that you choose for SourceList files isn't exactly the same as the one for Image files? I'm not 100% sure you'd have the subtraction/not subtraction in the filename. 

In my mind the SourceList is actually the same model for two very different datasets (just like Image) so making them obviously different will only help us. 

But I will have another look at the whole naming of these things and maybe I will change my mind.

Reply:
Since each source list comes from an image, ideally the source list filename will be the image filename with a different extension at the end (e.g. .sources.hdf5 or .sources.fits or whatever format we save it into).

I don't think of the source list datasets being all that different.  A subtraction image is different from a basic telescope image, I agree.  But, in both cases, the source list is just "what sources did I find on that image".  Deciding that they are new detections or not is really meaning that the user then ascribes to it-- if you are taking a deeper image than's been taken before of this field, you're going to have new detections in the raw image.   Detections and sources are really two different names for the same thing.